### PR TITLE
RUL-69: add attribute clearers

### PR DESCRIPTION
**Pull request by nmarniesse**
**Description (for Contributor and Core Developer)**

To clear an attribute value of a product, we had to call the property setter with the correct value (an empty string was working but this is not obvious for the option attributes for example).  
And clear product associations for example was painful because we had to build a specific format.  

To improve this part, we begin to implements property clearers. This is the first part with one clearer for all attribute types. 

This PR adds property clearer, with a registry and the first clearer for all atrribute types.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | OK
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Clearer/AttributeClearer.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function clear(EntityWithValuesInterface $entity, Attribute $attribute, array $options = []): void
+    {
+        $value = $entity->getValue($attribute->code(), $options['locale'] ?? null, $options['scope'] ?? null);
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Clearer/AttributeClearer.php]: why not directly pass the attribute code as string?

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Clearer/ClearerInterface.php [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Akeneo\Pim\Enrichment\Component\Product\Updater\Clearer;
+
+/**
+ * @author    Nicolas Marniesse <nicolas.marniesse@akeneo.com>
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface ClearerInterface
+{
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Clearer/ClearerInterface.php]: I'd put the `clear()` method here if I were you, and test the entity type in the implems 
(e.g for AttributeClearer, Assert::isInstanceOf($entity, EntityWithValues::class); in clear() method)

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/PropertyClearer.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        if ($clearer instanceof AttributeClearerInterface) {
+            $attribute = $this->getAttribute($property);
+            $clearer->clear($entity, $attribute, $options);
+        } elseif ($clearer instanceof FieldClearerInterface) {
+            $clearer->clear($entity, $property, $options);
+        } else {
+            throw UnknownPropertyException::unknownProperty($property);
+        }
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/PropertyClearer.php]: if you passed the attribute code as a string, and put the clear() method in the CleraraInterface, you could just replace these lines by
```php
    $clearer->clear(entitiy, $property, $options);
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Clearer/ClearerRegistryInterface.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    public function getAttributeClearer(Attribute $attribute): ?AttributeClearerInterface;
+
+    /**
+     * Get a field clearer compatible with given field.
+     *
+     * @param string $field
+     * @return FieldClearerInterface|null
+     */
+    public function getFieldClearer(string $field): ?FieldClearerInterface;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Clearer/ClearerRegistryInterface.php]: why are these methods `public`?

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Clearer/ClearerRegistry.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        $attribute = $this->getAttributes->forCode($property);
+
+        return null !== $attribute ? $this->getAttributeClearer($property) : $this->getFieldClearer($property);
+    }
+
+    private function getAttributeClearer(string $attributeCode): ?AttributeClearerInterface
+    {
+        foreach ($this->attributeClearers as $clearer) {
+            if ($clearer->supportsAttributeCode($attributeCode)) {
```
[the remark this answers is not in the document]
Comment (pierallard) [reply src/Akeneo/Pim/Enrichment/Component/Product/Updater/Clearer/ClearerRegistry.php]: Generally, I don't get why the meaning of the decoupling with all these interfaces.
Why just not a ClearerInterface, then a AttributeCleaner and a a Cleaner by attribute ? The AttributeCleaner should have the GetAttribute to check if this is an attribute or not.

**Pull request merged**